src/psyclone/domain/lfric/lfric_loop.py

```diff
@@ -232,48 +234,64 @@ def load(self, kern):
                 self.set_upper_bound("nannexed")
             else:
                 self.set_upper_bound("ndofs")
-        else:
+            return
+
+        if kern.iterates_over in ["halo_cell_column",
+                                  "owned_and_halo_cell_column"]:
             if Config.get().distributed_memory:
-                if self._field.is_operator:
-                    # We always compute operators redundantly out to the L1
-                    # halo
-                    self.set_upper_bound("cell_halo", index=1)
-                elif (self.field_space.orig_name in
-                      const.VALID_DISCONTINUOUS_NAMES):
-                    # Iterate to ncells for all discontinuous quantities,
-                    # including any_discontinuous_space
-                    self.set_upper_bound("ncells")
-                elif (self.field_space.orig_name in
-                      const.CONTINUOUS_FUNCTION_SPACES):
-                    # Must iterate out to L1 halo for continuous quantities
-                    # unless the only arguments that are updated all have
-                    # 'GH_WRITE' access. The only time such an access is
-                    # permitted for a field on a continuous space is when the
-                    # kernel is implemented such that any writes to a given
-                    # shared dof are guaranteed to write the same value. There
-                    # is therefore no need to iterate into the L1 halo in order
-                    # to get correct values for annexed dofs.
-                    if not kern.all_updates_are_writes:
-                        self.set_upper_bound("cell_halo", index=1)
-                    else:
-                        self.set_upper_bound("ncells")
-                elif (self.field_space.orig_name in
-                      const.VALID_ANY_SPACE_NAMES):
-                    # We don't know whether any_space is continuous or not
-                    # so we have to err on the side of caution and assume that
-                    # it is. Again, if the only arguments that are updated have
-                    # 'GH_WRITE' access then we can relax this condition.
-                    if not kern.all_updates_are_writes:
-                        self.set_upper_bound("cell_halo", index=1)
-                    else:
-                        self.set_upper_bound("ncells")
-                else:
-                    raise GenerationError(
-                        f"Unexpected function space found. Expecting one of "
-                        f"{const.VALID_FUNCTION_SPACES} but found "
-                        f"'{self.field_space.orig_name}'")
-            else:  # sequential
-                self.set_upper_bound("ncells")
+                if kern.iterates_over == "halo_cell_column":
+                    self.set_lower_bound("cell_halo_start")
+                self.set_upper_bound("cell_halo", index=kern.halo_depth)
+                return
```

Comment:
I have to admit that this is simpler than I thought it would be. I was expecting to have 4 loops, one for each halo, or a single loop with a masked body. But apparently it is enough with a single loop iterating over:
```
        assert "loop0_start = mesh%get_last_edge_cell()+1" in gen_code
        assert ("loop0_stop = mesh%get_last_halo_cell(hdepth)"
```
I suppose that this is because all the maps that index a "cell" to a position on the array have this values contiguous. Also because we start at "get_last_edge_cell()+1" we rely that the indexes are exactly after the owned ones, so:
"all mapping to all owned cells followed by all mapping to all hallo cells".

Is this the case or I am getting confused? Is this documented somewhere?

Also, is it safe to rely on this or is this just a implementation detail of the mapping that would break if somebody decided to reimplement the mapping infrastructure?

Reply:
As is often the way, in writing the email I realised that the diagram is in fact correct: the halo cell in the second partition is number '3' in the *local* indexing scheme so that it comes after the two owned cells.

Reply:
Ah yeah, something like that last sentence is what I wanted to see. Technically it should be extended to:
```
Notice the local indexing scheme is set up such that owned cells have lower indices than halo cells, the first halo cell starts immediately after the last owned cell, and each cell type are mapped contiguous indices.
```